File: Lab.SiteSPA/Scripts/Controllers/CarController.js
var CarListController = function ($http, $rootScope, $modal, $scope) {
    var ctrl = this;

    ctrl.cars = {};
    ctrl.isEmpty = false;

    ctrl.UserFunctions = $.grep($rootScope.permissions, function (e) { return e.currentFunction === "Cars" });

   $http({
        url: '/Car/GetCarList',
        method: 'POST' 
   }).success(function (result) {
       if (!$rootScope.accessDenied) {
           ctrl.cars = result;

           ctrl.isEmpty = ctrl.cars.length === 0 ? true : false;

           $rootScope.header = {
               showHeader: true,
               headerTitle: "Car List",
               hasAlert: ctrl.isEmpty,
               alertType: "warning",
               signType: "warning",
               message: "There are no cars to be displayed"
           };
       }
       $rootScope.accessDenied = false;       
    }).error(function () {
        if (!$rootScope.accessDenied) {
            $rootScope.header = {
                showHeader: true,
                headerTitle: "Car List",
                hasAlert: true,
                alertType: "danger",
                signType: "exclamation",
                message: "An Error has occured, the car list cannot be displayed"
            };
        }
        $rootScope.accessDenied = false;        
    });

    ctrl.openModal = function (id) {
        $rootScope.item = "car";
        var modalInstance = $modal.open({
            templateUrl: 'Scripts/Templates/DeleteModal.html',
            controller: 'ModalController',
            controllerAs: 'modalCtrl'
        });

        modalInstance.result.then(function (response) {
            $http({
                url: '/Car/DeleteCar',
                method: 'POST',
                params: {
                    id: id
                }
            }).success(function (result) {
                ctrl.cars = result;

                ctrl.isEmpty = ctrl.cars.length === 0 ? true : false;

                $rootScope.header = {
                    showHeader: true,
                    headerTitle: "Car List",
                    hasAlert: true,
                    alertType: "success",
                    signType: "ok",
                    message: "The car was deleted successfully"
                };
            }).error(function () {
                $rootScope.header = {
                    showHeader: true,
                    headerTitle: "Car List",
                    hasAlert: true,
                    alertType: "danger",
                    signType: "exclamation",
                    message: "An error has occured, the car could not be deleted"
                };
            });
        });
    };
};

var EditCarController = function ($http, $rootScope, $scope, $route, $routeParams) {
    var ctrl = this;

    ctrl.car = {
        Manufacturer: "",
        Model: "",
        Color: "",
        Year: null,
        Category: ""
    };

    ctrl.IsNew = true;

    ctrl.UserFunctions = $.grep($rootScope.permissions, function (e) { return e.currentFunction === "Cars" });

    if ($routeParams.Id) {
        $http({
            url: '/Car/GetCar',
            method: 'POST',
            params: {
                Id: $routeParams.Id
            }
        })
        .success(function (result) {
            if (!$rootScope.accessDenied) {
                ctrl.car = result;
                ctrl.IsNew = false;
                $rootScope.header = {
                    showHeader: true,
                    headerTitle: "Edit Car",
                    hasAlert: false,
                    alertType: "",
                    signType: "",
                    message: ""
                };
            }
            $rootScope.accessDenied = false;
        })
        .error(function () {
            if (!$rootScope.accessDenied) {
                $rootScope.header = {
                    showHeader: true,
                    headerTitle: "Edit Car",
                    hasAlert: true,
                    alertType: "danger",
                    signType: "exclamation",
                    message: "An error has occured, the car details cannot be displayed"
                };
            }
            $rootScope.accessDenied = false;
        });
    }
    else {
        $rootScope.header = {
            showHeader: true,
            headerTitle: "Add Car",
            hasAlert: false,
            alertType: "",
            signType: "",
            message: ""
        };
    }

    ctrl.saveCar = function (car) {
        var methodUrl = (ctrl.IsNew ? 'Car/CreateCar' : 'Car/UpdateCar');

        $http({ url: methodUrl, method: 'POST', params: { Data: JSON.stringify(car) } }).success(function (result) {
            if (!$rootScope.accessDenied) {
                ctrl.car = result;
                ctrl.IsNew = false;
                $rootScope.header = {
                    showHeader: true,
                    headerTitle: (ctrl.IsNew ? "Add" : "Edit") + " Car",
                    hasAlert: true,
                    alertType: "success",
                    signType: "ok",
                    message: "The car was saved succesfully"
                };
            }
            $rootScope.accessDenied = false;            
        }).error(function () {
            if (!$rootScope.accessDenied) {
                $rootScope.header = {
                    showHeader: true,
                    headerTitle: (ctrl.IsNew ? "Add" : "Edit") + " Car",
                    hasAlert: true,
                    alertType: "danger",
                    signType: "exclamation",
                    message: "An error has occured, the car could not be saved"
                };
            }
            $rootScope.accessDenied = false;            
        });
    };
};

File: MVCLab.Isthmus/Controllers/CarController.cs
using Lab.Business;
using Lab.Data.Model;
using Lab.MemberShip;
using MVCLab.Isthmus.Filters;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace MVCLab.Isthmus.Controllers
{
    public class CarController : Controller
    {
        [LabAuthorizeAttribute(Function = "Cars", Rights = "Read")]
        [RestoreModelStateFromTempData]
        public ActionResult CarList()
        {
            var cars = Cars.GetCarList();

            if (cars == null)
            {
                ViewBag.Message = "An Error has occured, the car list cannot be displayed";
                ViewData["MsgType"] = "Error";
            }

            if (cars.Count == 0)
            {
                ViewBag.Message = "There are no cars to be displayed";
                ViewData["MsgType"] = "Warning";
            }

            return View(cars);
        }

        [LabAuthorizeAttribute(Function = "Cars", Rights = "Create")]
        public ViewResult AddCar()
        {
            var car = new Car();

            return View("EditCar", car);
        }

        [LabAuthorizeAttribute(Function = "Cars", Rights = "Create")]
        [SetTempDataModelState]
        public ActionResult CreateCar(Car car, string returnUrl)
        {
            var Car = Cars.SaveCar(car);

            if (Car == null)
            {
                ViewBag.Message = "An error has occured, the car could not be saved";
                ViewData["MsgType"] = "Error";

                return View("EditCar", car);
            }

            ViewData["MsgType"] = "Success";
            ViewBag.Message = "The car was saved successfully";

            return RedirectToAction("EditCar", "Car", new { id = Car.Id });
        }

        [LabAuthorizeAttribute(Function = "Cars", Rights = "Read")]
        [RestoreModelStateFromTempData]
        public ViewResult EditCar(int Id)
        {
            var car = Cars.GetCarById(Id);

            if (car == null)
            {
                ViewBag.Message = "An error has occured, the car details cannot be displayed";
                ViewData["MsgType"] = "Error";
            }

            return View(car);
        }

        [LabAuthorizeAttribute(Function = "Cars", Rights = "Update")]
        [SetTempDataModelState]
        public ActionResult UpdateCar(Car car, string returnUrl)
        {
            var Car = Cars.SaveCar(car);

            if (Car == null)
            {
                ViewBag.Message = "An error has occured, the car could not be saved";
                ViewData["MsgType"] = "Error";

                return View("EditCar", car);
            }

            ViewData["MsgType"] = "Success";
            ViewBag.Message = "The car was saved successfully";

            return RedirectToAction("EditCar", "Car", new { id = Car.Id });
        }

        [LabAuthorizeAttribute(Function = "Cars", Rights = "Delete")]
        [SetTempDataModelState]
        public ActionResult DeleteCar(int id)
        {
            var result = Cars.DeleteCar(id);

            if (!result)
            {
                ViewBag.Message = "An error has occured, the car could not be deleted";
                ViewData["MsgType"] = "Error";
            }

            ViewData["MsgType"] = "Success";
            ViewBag.Message = "The car was deleted successfully";

            return RedirectToAction("CarList", "Car");
        }
    }
}
